### PR TITLE
JBPM-7912 text annotation

### PR DESCRIPTION
**Pull request by tiagodolphine**
Added the text annotation feature.

- Shape on canvas
- Icon on the palette
- Quick menu to add text annotation
- Created association connector without direction (the arrow to the target node)
- Marshalling to support text annotation in the same way as the jbpm desginer outputs

@romartin 
@LuboTerifaj

### CODE REVIEW

**Comment by tiagodolphine** [comment]
jenkins execute full downstream build

**Comment by romartin** [comment]
ooops conflicts :)

**Comment by tiagodolphine** [comment]
jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
@romartin conflicts solved

**Comment by tiagodolphine** [comment]
@romartin @LuboTerifaj when you have a chance to take a look on this one, mainly doing some manual tests to see how it behaves on canvas, palette, quick menu, etc.

**Comment by hasys** [comment]
Hi @romartin, I think it is approach of the new Designer's palette to create new global menus almost for everything for quick access. So in my opinion it is OK as it is right now.

But final word should be on UX team. @lclay2 what do you think?

**Comment by krisv** [comment]
I guess the category should probably be "artefacts".  There are different types of artefacts, text annotation would be one of them.  Others are data objects / inputs / outputs and groups, but since we don't implement any of them yet, I think it's fine if for now only text annotation shows up in there.

**Comment by ibek** [comment]
I would add Text Annotations under Tasks. I think it's not worth to have another group because it will be taking more place on the screen and anytime you need to add text-annotation you have to switch the palette. Moreover, there's nothing else we would be adding into that group AFAIK ... I think that if you start modeling your process with tasks you add annotation as a reminder or to explain some parts as you add the tasks ... that's why I think Text Annotation would fit in there nicely. Please ask also Liz C.

**Comment by ibek** [comment]
Kris, true, I forgot about the data objects and data sources.

**Comment by lclay2** [comment]
@ibek @krisv As you know the design "pattern" is to group like things, to keep the palette both ordered and manageable. The downside to the palette groupings is if the high level category is not obvious, they the user might not know to look in that grouping for an item, and might have to hunt a bit at first to find and learn it's in there. It could be argued that this is an acceptable trade-off for lower priority or infrequently used elements - although for commonly used elements it could pose a usability problem. So for this reason it's best to avoid "misc" (junk drawer) category grouping. If the user associates Text Annotations with Task or Artifacts, then that seems fine to group it this way. If not, or if we're not sure, then imo I would err on the side of making it obvious and leaving it at the top level.

**Comment by tiagodolphine** [comment]
Ok so, As I understood, let's keep the way it is now, right?
@romartin @LuboTerifaj let's proceed with the review then.

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/DirectionalAssociation.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.stunner.bpmn.definition;
+
+/*
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/DirectionalAssociation.java]: done

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java [hunk 2]
```diff
@@ -2359,15 +2359,38 @@ public void testSignalRefEqualsAndHashCode() {
...
+                .addFalseCase(new DirectionalAssociation(),
+                              new DirectionalAssociation(new BPMNGeneralSet(), new BackgroundSet(), new FontSet()))
+                .addFalseCase(new DirectionalAssociation(),
+                              new DirectionalAssociation(null, null, null))
+                .test();
+    }
+
+    @Test
+    public void testNonDirectionalAssociationEqualsAndHashCode() {
+        TestCaseBuilder.newTestCase()
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java]: done

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/associations/AssociationPropertyWriterTest.java [hunk 3]
```diff
@@ -114,7 +116,17 @@ public void testSetTarget() {
...
+    @Test
     public void testSetOneDirectionAssociation() {
+        DirectionalAssociation directionalAssociation = new DirectionalAssociation();
+        associationWriter.setDirectionAssociation(directionalAssociation);
         verify(association).setAssociationDirection(AssociationDirection.ONE);
     }
+
+    @Test
+    public void testSetNoneDirectionAssociation() {
+        NonDirectionalAssociation directionalAssociation = new NonDirectionalAssociation();
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/associations/AssociationPropertyWriterTest.java]: done

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/AssociationPropertyReaderTest.java [hunk 4]
```diff
@@ -123,6 +127,24 @@ public void testGetTargetConnection() {
         assertEquals(Y, result.getLocation().getY(), 0);
     }
 
+    @Test
+    public void testGetAssociationByDirection(){
```
[the remark this answers is not in the document]
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/AssociationPropertyReaderTest.java]: done

**Comment by LuboTerifaj** [comment]
Hello @tiagodolphine ,

I have found two more issues:

- Regarding bend-points, if you add a bend-point to non directional association inside sub-process, after moving with sub-process to another position, it seems that the bend-point disappears. Actually, it stays there, but the association is not bended any more. It's straight.
- Second one is related to process migration.
If you create any node with Non Directional Association connected to Text Annotation in legacy designer, after the process migration and moving with any node, it is not possible to save the process with the message:
`The connection is not allowed for the edge with role 'org.kie.workbench.common.stunner.bpmn.definition.NonDirectionalAssociation'. The permitted connections are '[{'fromtoall' ->'text_annotation'}]'`

Thanks!

**Comment by tiagodolphine** [comment]
jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Solved some comments ande the bellow issues:
- It is not possible to save process with node inside Lane and connected Text Annotation to this node, which is outside of the Lane.
- It is not possible to add Text Annotation to the Lane and Ad-hoc Sub-process
It is not possible to connect any node with Text Annotation between different levels (top-level <-> sub-process). It should be possible to let being Text Annotation created everywhere. **This one is partially solved, it is not working with sub-processes only Lanes.**

Analyzing the other issues.

**Comment by romartin** [comment]
Thanks @tiagodolphine for the great work! 

@tiagodolphine  @LuboTerifaj Plz guys as commented on last checkpoint meeting let's try to be "pragmatic", as today is the last day for including new features targeting 7.5, let's just address blocker issues and report other issues as separate tickets, if possible.

Anyway great job both, thanks a lot! :+1:

**Comment by tiagodolphine** [comment]
It fails on an unrelated test error and the WARs were generated.

**Comment by tiagodolphine** [comment]
@romartin @LuboTerifaj yeah I agree, we need to see if the remaining issues can be opened as non blocker bugs and we work during today and next week, I mean one of them seem to be more complex to solve because it requires a change on the flow used on the marshallers, regarding sub-processes and association with node out of the sub-process (this may cause the issue with text-annotation on top-level -> association -> subprocess)
And regarding the bendpoints stuff I didn't change anything from this perspective, so in opinion is that thease issues are not related/blocker and we can open a new JIRA for it
The last thing regarding case view switched to process view, well I don't know how this works to fix quickly, but I think it is a non blocker as we can open a new JIRA as well.

WDYT guys?

**Comment by romartin** [comment]
ok thanks @LuboTerifaj for the great review.

Let's try to address the blocker issue anyway ASAP plz @tiagodolphine , so will see if we can introduce this commit at some point... thx!

**Comment by tiagodolphine** [comment]
jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Hi @LuboTerifaj the fix I commented today on the call worked, I built and tested the WB.
I saw you approved, but can you give the last review, in this way we consider the bug with Case Management fixed.
cc @romartin 
Thanks

**Comment by romartin** [comment]
ok thanks @LuboTerifaj 
Please @tiagodolphine lemme know once builds are ready and I'll merge! 
Thanks guys!

**Comment by LuboTerifaj** [comment]
Hello @tiagodolphine , @romartin ,

The issues mentioned above are already reported:

- [JBPM-8719](https://issues.jboss.org/browse/JBPM-8719) Stunner - [Case Modeler] Not possible to save case containing Text Annotation
- [JBPM-8720](https://issues.jboss.org/browse/JBPM-8720) Stunner - Non-directional association bend-points are placed still on canvas
- [JBPM-8721](https://issues.jboss.org/browse/JBPM-8721) Stunner - Non-directional association is not persisted between containers
- [JBPM-8723](https://issues.jboss.org/browse/JBPM-8723) Stunner - Bend-points for Non-directional Association don't work inside sub-process
- [JBPM-8724](https://issues.jboss.org/browse/JBPM-8724) Stunner - Make bend-points for Non-directional Association creation easier

Thanks!

**Comment by tiagodolphine** [comment]
Hi @LuboTerifaj I think the Case Modeler issue is fixed, at least I pushed a commit yesterday to fix it. Since this is already merged on the master, it is probably already there, I don't know if you are able to download the last WAR from this PR, but it would be great if you can validate it.

**Comment by LuboTerifaj** [comment]
Hi @tiagodolphine ,

I was testing the newest snapshot with this PR already merged. All issues above are present there.

Thanks!

**Pull request merged**